sandbox/apparmor/process.go

```diff
@@ -0,0 +1,68 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package apparmor
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func labelFromPid(pid int) (string, error) {
+	procFile := filepath.Join(rootPath, fmt.Sprintf("proc/%v/attr/current", pid))
+	contents, err := ioutil.ReadFile(procFile)
+	if os.IsNotExist(err) {
+		return "unconfined", nil
+	} else if err != nil {
+		return "", err
+	}
+	label := strings.TrimRight(string(contents), "\n")
+	// Trim off the mode
+	if strings.HasSuffix(label, ")") {
+		if pos := strings.LastIndex(label, " ("); pos != -1 {
+			label = label[:pos]
+		}
+	}
+	return label, nil
+}
+
+func decodeLabel(label string) (snap, app, hook string, err error) {
```

Comment:
I've not looked at this PR in depth, but I have no real objections to it since this seems to be taking existing xdg-desktop-portal code out of that project and putting it into this one. As such, there is no real regression risk, etc.

That said, I guess we aren't using the libapparmor API (eg, aa_getcon()) cause, well, snapd and libapparmor is C. It would be a bit more futureproof if we did, but again, I'm not blocking. I can say that this code is not considering apparmor namespaces. Eg, consider this label pulled from `/proc/<container pid>/attr/current` from the host on a snap process running in an lxd container: `lxd-sb_</var/snap/lxd/common/lxd>//&:lxd-sb_<var-snap-lxd-common-lxd>:snap.hello-world.sh`. To be fair, the snapd managing that process is only going to see `snap.hello-world.sh` (ie, from in the container, `/proc/<pid>/attr/current` will show what we want) *and* we don't want the snapd on the host to manage the processes of the snapd in the container, so this is also not a blocker IME. It does underscore that there are other things out there though.

All of this is a roundabout way of saying that I think we need a code comment. Perhaps:
```
// XXX: we aren't handling apparmor namespaces properly here. In practice,
// snapd only considers profiles in its (relative) namespace that it owns, and
// not those in snapd's in other namespaces (eg, snapd in lxd) so this isn't
// currently an issue.
```

Reply:
If we end up being able to get application information from cgroups, then we can eventually remove this AppArmor code.  Although I guess the same kind of cross-container shenanigans could be an issue there too.

I'm not sure whether sharing a D-Bus session bus across container boundaries is a great idea though...